sdk/storage/Azure.Storage.Internal.Avro/src/AvroParser.cs

```diff
@@ -0,0 +1,388 @@
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadFixedBytesAsync(stream, 1, async, cancellationToken).ConfigureAwait(false);
+            return bytes[0];
+        }
+
+        // Stolen because the linked references in the Avro spec were subpar...
+        private static async Task<long> ReadZigZagLongAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            ulong next = b & 0x7FUL;
+            int shift = 7;
+            while ((b & 0x80) != 0)
+            {
+                b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+                next |= (b & 0x7FUL) << shift;
+                shift += 7;
+            }
+            long value = (long)next;
+            return (-(value & 0x01L)) ^ ((value >> 1) & 0x7fffffffffffffffL);
+        }
+
+        public static Task<object> ReadNullAsync() => Task.FromResult<object>(null);
+
+        public static async Task<bool> ReadBoolAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte b = await ReadByteAsync(stream, async, cancellationToken).ConfigureAwait(false);
+            return b != 0;
+        }
+
+        public static async Task<long> ReadLongAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken) =>
+            await ReadZigZagLongAsync(stream, async, cancellationToken).ConfigureAwait(false);
+
+        public static async Task<int> ReadIntAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken) =>
+            (int)(await ReadLongAsync(stream, async, cancellationToken).ConfigureAwait(false));
+
+        public static async Task<float> ReadFloatAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadFixedBytesAsync(stream, 4, async, cancellationToken).ConfigureAwait(false);
+            return BitConverter.ToSingle(bytes, 0);
+        }
+
+        public static async Task<double> ReadDoubleAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            byte[] bytes = await ReadFixedBytesAsync(stream, 8, async, cancellationToken).ConfigureAwait(false);
+            return BitConverter.ToDouble(bytes, 0);
+        }
+
+        public static async Task<byte[]> ReadBytesAsync(
+            Stream stream,
+            bool async,
+            CancellationToken cancellationToken)
+        {
+            int size = await ReadIntAsync(stream, async, cancellationToken).ConfigureAwait(false);
```

Comment:
can you either leave a comment saying this should be long as per spec, use int since array length is int or just make this read a long and convert that to int explicitly?

Reply:
Added a comment.